openedx/core/djangoapps/util/row_delete.py

```diff
@@ -0,0 +1,102 @@
+"""
+Code to delete rows from a table within a Django mgmt command using best practices.
+"""
+
+import logging
+import time
+from django.core.management.base import CommandError, BaseCommand
+from django.db import transaction
+
+log = logging.getLogger(__name__)
+
+
+def delete_rows(model_mgr,
+                table_name,
+                primary_id_name,
+                chunk_size,
+                sleep_between):
+    """
+    Deletes *ALL* rows from table, chunking the deletes to avoid long table/row locks.
+
+    Args:
+        model_mgr (django.db.models.manager.Manager): Django ORM mgr for the table's model.
+        table_name (str): Name of table from which to delete all rows.
+        primary_id_name (str): Name of primary ID autoincrement column from table.
+        chunk_size (int): Number of rows to delete in each transaction.
+        sleep_between (float): Number of seconds to sleep between transactions.
+    """
+    if chunk_size <= 0:
+        raise CommandError('Only positive chunk size is allowed ({}).'.format(chunk_size))
+    if sleep_between < 0:
+        raise CommandError('Only non-negative sleep between seconds is allowed ({}).'.format(sleep_between))
+
+    # The "as id" below fools Django raw query into thinking the primary key is being queried.
+    min_max_ids = model_mgr.raw(
+        'SELECT MIN({}) as id, MAX({}) as max_id FROM {}'.format(primary_id_name, primary_id_name, table_name)
+    )[0]
+    min_id = min_max_ids.id
+    max_id = min_max_ids.max_id
+    if not min_id or not max_id:
+        log.info("No data exists in table %s - skipping.", table_name)
+        return
+    log.info(
+        "STARTED: Deleting around %s rows with chunk size of %s and %s seconds between chunks.",
+        max_id - min_id + 1, chunk_size, sleep_between
+    )
+
+    total_deletions = 0
+    lower_id = min_id
+    while lower_id <= max_id:
+        deletions_now = min(chunk_size, max_id - lower_id + 1)
+        upper_id = lower_id + deletions_now
+        log.info("Deleting around %s rows between ids %s and %s...", deletions_now, lower_id, upper_id)
+        with transaction.atomic():
+            delete_sql = 'DELETE FROM {} WHERE {} >= {} AND {} < {}'.format(
```

Comment:
Why not use the ORM's delete function instead?

Reply:
The primary reason: The django-simple-history module will be removed before these commands are run - in order to shut off the recording of history. So the model will no longer know how to access the history rows.
The secondary reason: The HistoricalRecords() field creates a model dynamically. Although the docs talk of querying history records via `.history`, it mentions nothing about record deletion, so I wasn't sure if it was possible/tested/likely to work.